src/libsyntax/ast.rs

```diff
@@ -2106,20 +2106,54 @@ pub struct AttrItem {
 }
 
 /// Metadata associated with an item.
-/// Doc-comments are promoted to attributes that have `is_sugared_doc = true`.
 #[derive(Clone, RustcEncodable, RustcDecodable, Debug)]
 pub struct Attribute {
-    pub item: AttrItem,
+    pub kind: AttrKind,
     pub id: AttrId,
     pub style: AttrStyle,
-    pub is_sugared_doc: bool,
     pub span: Span,
 }
 
-// Compatibility impl to avoid churn, consider removing.
-impl std::ops::Deref for Attribute {
-    type Target = AttrItem;
-    fn deref(&self) -> &Self::Target { &self.item }
+#[derive(Clone, RustcEncodable, RustcDecodable, Debug)]
+pub enum AttrKind {
+    /// The normal representation.
+    Normal(AttrItem),
+
+    /// A specialized representation for doc comments that exists purely as an
+    /// optimization. `self.has_name(sym::doc)` and `self.check_name(sym::doc)`
+    /// succeed for this variant.
+    ///
+    /// `SugaredDoc(comment)` is equivalent to the following, with the `_`
+    /// parts left unspecified:
+    /// ```
+    /// AttrKind::Normal(
+    ///     AttrItem {
+    ///         path: Path {
+    ///             span: _,
+    ///             segments: vec![
+    ///                 PathSegment {
+    ///                     ident: Ident { name: sym::doc, span: _ },
+    ///                     id: _,
+    ///                     args: None
+    ///                 }
+    ///             ]
+    ///         },
+    ///         tokens: TokenStream(
+    ///             Lrc(vec![
+    ///                 TokenTree::Token(Token {
+    ///                     kind: Eq,
+    ///                     span: _
+    ///                 }),
+    ///                 TokenTree::Token(Token {
+    ///                     kind: Literal(Lit { kind: Str, symbol: comment, suffix: None }),
+    ///                     span: _
+    ///                 })
+    ///             ])
+    ///         )
+    ///     }
+    /// )
+    /// ```
+    SugaredDoc(Symbol),
```

Comment:
```suggestion
    DocComment(Symbol),
```
Could you also mass rename the remaining `SugaredDoc`/`sugared_doc`s into `DocComment`/`doc_comment`s?

Reply:
I think `SugaredDoc` is better because it clearly only matches sugared doc comments, e.g. `///`. `DocComment` could also mean `#[doc="..."]` comments.

Reply:
`#[doc]` is not a comment, it's a regular attribute.
Yes, there are two ways to drive rustdoc - doc comments and `#[doc]` attributes, but that certainly doesn't mean `#[doc]` attributes are comments.
`/// Doc` being a comment and `#[doc]` not being a comment is exactly the point of using the name `DocComment`.

So, `DocComment` literally names the thing it denotes, while `SugaredDoc` doesn't, it's like naming `While` `SugaredLoopWithConditionalBreak` or something.

Reply:
So, I disagree and this mixing of doc comments and doc attributes in the frontend is exactly what I'd like to eliminate (or at least minimize).